vscode/src/autoedits/adapters/fireworks-websocket.ts

```diff
@@ -0,0 +1,197 @@
+        const messageId = 'm_' + this.messageId++
+        const data = JSON.stringify({
+            'x-message-id': messageId,
+            'x-message-body': body,
+            'x-message-url': url,
+            'x-message-headers': headers,
+        })
+
+        if (messageId in this.callbackQueue) {
+            autoeditsOutputChannelLogger.logError(
+                LOG_FILTER_LABEL,
+                `unexpected, duplicate message ID ${messageId}`
+            )
+        }
+
+        const sendPromise: Promise<Response> = new Promise(resolve => {
+            this.callbackQueue[messageId] = resolve
+            this.ws?.send(data)
+        })
+
+        let timeoutHandle: NodeJS.Timeout
+        const timeoutPromise: Promise<Response> = new Promise((_, reject) => {
+            timeoutHandle = setTimeout(() => {
+                delete this.callbackQueue[messageId]
+                reject(new Error('timeout in sending message to fireworks'))
+            }, SOCKET_SEND_TIME_OUT_MS)
+        })
+
+        return Promise.race([sendPromise, timeoutPromise]).then(response => {
+            clearTimeout(timeoutHandle)
+            return response
+        })
+    }
+
+    private reconnect() {
+        ;(async () => {
+            await this.connect()
+        })()
+    }
+
+    private async connect(): Promise<WebSocket> {
+        return new Promise((resolve, reject) => {
+            if (this.ws && this.ws.readyState === WebSocket.OPEN) {
+                resolve(this.ws)
+                return
+            }
+
+            const ws = new WebSocket(this.webSocketEndpoint)
+            ws.addEventListener('open', () => {
+                autoeditsOutputChannelLogger.logDebug(
+                    LOG_FILTER_LABEL,
+                    `successfully connected to ${this.webSocketEndpoint}`
+                )
+                this.ws = ws
+                resolve(this.ws!)
+            })
+            ws.addEventListener('error', (event: ErrorEvent) => {
+                autoeditsOutputChannelLogger.logError(
+                    LOG_FILTER_LABEL,
+                    `error from ${this.webSocketEndpoint}: ${event.message}`
+                )
+                if (process.env.NODE_ENV === 'development') {
+                    console.error(`error from ${this.webSocketEndpoint}: ${event.message}`)
+                    console.error(event)
+                }
+                reject(event)
+            })
+            ws.addEventListener('close', (event: CloseEvent) => {
+                autoeditsOutputChannelLogger.logDebug(
+                    LOG_FILTER_LABEL,
+                    `${this.webSocketEndpoint} connection closed with code ${event.code}`
+                )
+                if (process.env.NODE_ENV === 'development') {
+                    console.error(`${this.webSocketEndpoint} connection closed`)
+                    console.error(event)
+                }
+                setTimeout(this.reconnect, SOCKET_RECONNECT_DELAY_MS)
```

Comment:
To keep `this` context for the `this.reconnect` call we can do the following:

```suggestion
                setTimeout(() => this.reconnect(), SOCKET_RECONNECT_DELAY_MS)
                
```

Or convert the `this.reconnect` method to an arrow function.

Reply:
Thoughts for the follow-up: We probably want an exponential backoff here for cases where the API is not available to avoid spamming the CPU with requests to reconnect.